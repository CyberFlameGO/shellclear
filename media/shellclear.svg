<svg xmlns="http://www.w3.org/2000/svg" width="260" height="260" fill="none" viewBox="0 0 260 260"><rect width="208" height="168" x="26" y="46" fill="#000" stroke="#fff" stroke-width="8" rx="21"/><path fill="#EBEBEB" stroke="#000" d="M66.8746 82.7701L66.8653 82.7615L66.8556 82.7534L54.2904 72.2673C50.7899 69.3461 45.5 71.9268 45.5 76.4558C45.5 78.0917 46.2738 79.6507 47.5906 80.628L58.8592 88.9915V89.0094L47.6189 97.3752C46.2849 98.3681 45.5 99.9411 45.5 101.602C45.5 106.131 50.7868 108.657 54.2754 105.745L66.8556 95.2466L66.5352 94.8627L66.8675 95.2363C70.5517 91.9595 70.6067 86.2197 66.9858 82.8729L66.8746 82.7701Z"/><rect width="66" height="17" x="146.5" y="113.5" fill="#F33" stroke="#000"/><rect width="83" height="17" x="89.5" y="81.5" fill="#F33" stroke="#000"/><rect width="20" height="17" x="192.5" y="81.5" fill="#F33" stroke="#000"/><rect width="72" height="17" x="46.5" y="175.5" fill="#F33" stroke="#000"/><rect width="46" height="7" x="46.5" y="118.5" fill="#8A8A8A" stroke="#000"/><rect width="10" height="7" x="78.5" y="86.5" fill="#8A8A8A" stroke="#000"/><rect width="37" height="7" x="150.5" y="148.5" fill="#8A8A8A" stroke="#000"/><rect width="24" height="7" x="119.5" y="180.5" fill="#8A8A8A" stroke="#000"/><rect width="46" height="7" x="99.5" y="118.5" fill="#8A8A8A" stroke="#000"/><rect width="8" height="7" x="194.5" y="148.5" fill="#8A8A8A" stroke="#000"/><rect width="12" height="7" x="173.5" y="86.5" fill="#8A8A8A" stroke="#000"/><rect width="4" height="7" x="123.5" y="148.5" fill="#8A8A8A" stroke="#000"/><rect width="10" height="7" x="67.5" y="148.5" fill="#8A8A8A" stroke="#000"/><rect width="18" height="7" x="150.5" y="180.5" fill="#8A8A8A" stroke="#000"/><rect width="20" height="17" x="46.5" y="143.5" fill="#F33" stroke="#000"/><rect width="38.242" height="17" x="78.379" y="143.5" fill="#F33" stroke="#000"/><rect width="37" height="17" x="128.5" y="143.5" fill="#F33" stroke="#000"/></svg>