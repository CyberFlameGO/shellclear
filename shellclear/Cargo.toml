[package]
name = "shellclear"
version = "0.2.2"
edition = "2021"
authors = ["Elad-Kaplan <kaplan.elad@gmail.com>"]
readme = "README.md"
repository = "https://github.com/rusty-ferris-club/shellclear"
homepage = "https://github.com/rusty-ferris-club/shellclear"
build = "build.rs"

[dependencies]
clap = { version = "3", features = ["cargo"], optional = true }
serde = "1"
serde_derive = "1"
serde_yaml = "^0.8.24"
serde_regex = "1.1.0"
env_logger = "^0.9.0"
log = "^0.4.17"
anyhow = "1"
console = "^0.15.0"
home = "0.5.3"
rayon = "1.5"
regex = "1"
chrono = "0.4"
strum = { version = "0.24", features = ["derive"] }
prettytable-rs = "^0.8"
dialoguer = "^0.10.1"
ctrlc = "3.2.2"
exitcode = "^1.1.2"

[dev-dependencies]
insta = { version = "1", features = ["backtrace"] }
tempdir = "0.3"

[build-dependencies]
vergen = "7"
anyhow = "1"


[features]
default = ["cli"]
# list optionals here:
cli = ["clap"]

[[bin]]
name = "shellclear"
required-features = ["cli"]
