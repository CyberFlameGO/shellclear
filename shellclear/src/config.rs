use crate::data::{SensitiveCommands, ROOT_APP_FOLDER};
use anyhow::Result;
use std::fs;
use std::path::PathBuf;

const CONFIG_SENSITIVE_PATTERNS: &str = "sensitive-patterns.yaml";
const SENSITIVE_PATTERN_TEMPLATE: &str = r###"# External sensitive patters file allows you you add a custom patterns to shellclear

- name: Pattern Name
  test: <PATTERN REGEX>
"###;

#[derive(Clone, Debug)]
pub struct Config {
    sensitive_commands_path: PathBuf,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            sensitive_commands_path: Config::get_sensitive_pattern_file(&dirs::home_dir().unwrap()),
        }
    }
}

impl Config {
    fn with_custom_path(root: &PathBuf) -> Self {
        Self {
            sensitive_commands_path: Self::get_sensitive_pattern_file(root),
        }
    }
    /// Init external configuration
    ///
    /// # Errors
    ///
    /// Will return `Err` when home directory not found or failed to create a file
    pub fn init(&self) -> Result<String> {
        let file = &self.sensitive_commands_path;
        fs::write(file, SENSITIVE_PATTERN_TEMPLATE)?;
        Ok(file.display().to_string())
    }

    /// Get sensitive file path name
    ///
    /// # Errors
    ///
    /// Will return `Err` home directory not found
    #[must_use]
    pub fn get_sensitive_pattern_name(&self) -> String {
        self.sensitive_commands_path.display().to_string()
    }

    /// Is sensitive file is exists
    ///
    /// # Errors
    ///
    /// Will return `Err` home directory not found
    #[must_use]
    pub fn is_sensitive_pattern_file_exists(&self) -> bool {
        self.sensitive_commands_path.exists()
    }

    /// Load sensitive pattern file
    ///
    /// # Errors
    ///
    /// Will return `Err` home directory not found or yaml is invalid
    pub fn load_patterns_from_default_path(&self) -> Result<Vec<SensitiveCommands>> {
        self.load_patterns_from_file(&self.sensitive_commands_path)
    }

    /// Load sensitive pattern from the given path
    ///
    /// # Errors
    ///
    /// Will return `Err`  yaml is invalid
    fn load_patterns_from_file(&self, path: &PathBuf) -> Result<Vec<SensitiveCommands>> {
        let f = std::fs::File::open(path)?;
        Ok(serde_yaml::from_reader(f)?)
    }

    /// Returns the root shellclear config folder
    fn get_sensitive_pattern_file(path: &PathBuf) -> PathBuf {
        path.join(format!(".{}", ROOT_APP_FOLDER))
            .join(CONFIG_SENSITIVE_PATTERNS)
    }
}
