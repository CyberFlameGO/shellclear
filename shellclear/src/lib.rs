mod data;
mod emoji;
mod shell;
mod state;

pub mod config;
pub mod dialog;
pub mod engine;
pub mod exporter;
pub use self::data::{CmdExit, FindingSensitiveCommands, SensitiveCommands};
pub use self::emoji::Emojis;
pub use self::state::{init, ShellContext};
