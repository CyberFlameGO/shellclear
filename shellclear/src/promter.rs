use anyhow::anyhow;
use anyhow::Result;
use dialoguer::{theme::ColorfulTheme, Confirm, Select};

/// prompt select option
///
/// # Errors
///
/// Will return `Err` when interact error
pub fn select(message: &str, items: &Vec<String>) -> Result<usize> {
    let selection = match items.len() {
        1 => 0,
        _ => match Select::with_theme(&ColorfulTheme::default())
            .with_prompt(message)
            .default(0)
            .items(items)
            .interact()
        {
            Ok(s) => s,
            Err(e) => return Err(anyhow!("{}", e)),
        },
    };

    Ok(selection)
}

/// prompt confirm message
///
/// # Errors
///
/// Will return `Err` when interact error
pub fn confirm(message: &str) -> Result<()> {
    match Confirm::with_theme(&ColorfulTheme::default())
        .with_prompt(message)
        .interact()
    {
        Ok(is_confirm) => {
            if !is_confirm {
                return Err(anyhow!("not confirmed"));
            }
        }
        Err(e) => {
            log::debug!("confirm interact err: {}", e);
            return Err(anyhow!("confirm interact err: {}", e));
        }
    }
    Ok(())
}
