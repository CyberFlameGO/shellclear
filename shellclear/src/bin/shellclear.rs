mod cmd;
use anyhow::anyhow;
use console::style;
use shellclear::{init, promter, ShellContext};
use std::process::exit;

const DEFAULT_ERR_EXIT_CODE: i32 = 1;

pub const BANNER: &str = r#"
     _          _ _      _                 
 ___| |__   ___| | | ___| | ___  __ _ _ __ 
/ __| '_ \ / _ \ | |/ __| |/ _ \/ _` | '__|
\__ \ | | |  __/ | | (__| |  __/ (_| | |   
|___/_| |_|\___|_|_|\___|_|\___|\__,_|_| "#;

fn main() {
    let app = cmd::default::command()
        .subcommand(cmd::find::command())
        .subcommand(cmd::restore::command())
        .subcommand(cmd::stash::command());

    let matches = app.clone().get_matches();

    let env = env_logger::Env::default().filter_or(
        "LOG",
        matches.value_of("log").unwrap_or(log::Level::Info.as_str()),
    );
    env_logger::init_from_env(env);

    if !matches.is_present("no_banner") {
        println!(
            "{}{}",
            style(BANNER).magenta(),
            style(app.get_version().unwrap_or("")).dim()
        );
    }

    if let Err(e) = ctrlc::set_handler(move || {
        let term = console::Term::stdout();
        let _ = term.show_cursor();
    }) {
        log::debug!("{:?}", e);
    }

    // create app config to store state data
    let shells_context = match init() {
        Ok(s) => s,
        Err(e) => {
            log::error!("{}", e);
            exit(1)
        }
    };

    let res = match matches.subcommand() {
        None => Err(anyhow!("command not found")),
        Some(tup) => match tup {
            ("find", subcommand_matches) => cmd::find::run(subcommand_matches, &shells_context),
            ("restore", _subcommand_matches) => cmd::restore::run(select_shell(&shells_context)),
            ("stash", subcommand_matches) => {
                cmd::stash::run(subcommand_matches, select_shell(&shells_context))
            }
            _ => unreachable!(),
        },
    };

    let exit_with = match res {
        Ok(cmd) => {
            if let Some(message) = cmd.message {
                eprintln!("{}", message);
            }
            cmd.code
        }
        Err(e) => {
            log::debug!("{:?}", e);
            DEFAULT_ERR_EXIT_CODE
        }
    };
    exit(exit_with)
}

fn select_shell(shell_contexts: &Vec<ShellContext>) -> &ShellContext {
    if shell_contexts.len() == 1 {
        return &shell_contexts[0];
    }
    let selections = shell_contexts
        .iter()
        .map(|f| format!("{:?}     : {}", f.history.shell, f.history.path))
        .collect::<Vec<_>>();

    match promter::select("Pick your shell", &selections) {
        Ok(selection) => &shell_contexts[selection],
        Err(e) => {
            log::debug!("promter select err: {:?} ", e);
            exit(1)
        }
    }
}
