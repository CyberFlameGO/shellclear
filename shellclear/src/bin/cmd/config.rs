use anyhow::Result;
use clap::{Arg, ArgMatches, Command};
use shellclear::config::Config;
use shellclear::promter;

pub fn command() -> Command<'static> {
    Command::new("config")
        .about("Create custom configuration")
        .arg(
            Arg::new("validate")
                .short('v')
                .long("validate")
                .help("Validate configuration file.")
                .takes_value(false),
        )
        .arg(
            Arg::new("delete")
                .short('d')
                .long("delete")
                .help("Delete configuration file.")
                .takes_value(false),
        )
}

pub fn run(matches: &ArgMatches, config: &Config) -> Result<shellclear::CmdExit> {
    let file_path = config.get_sensitive_pattern_name();

    let (message, exit_code) = if matches.is_present("validate") {
        if let Err(e) = config.load_patterns_from_default_path() {
            (
                format!("Config file is invalid. error: `{}`.", e),
                exitcode::CONFIG,
            )
        } else {
            (
                format!("configuration file: {} is valid.", file_path),
                exitcode::OK,
            )
        }
    } else if matches.is_present("delete") {
        if let Err(e) = promter::confirm(format!("Delete {} file?", &file_path).as_str()) {
            log::debug!("{:?}", e);
            return Ok(shellclear::CmdExit {
                code: exitcode::OK,
                message: None,
            });
        }

        if config.is_sensitive_pattern_file_exists() {
            config.delete_sensitive_patterns_from_file()?;
        }

        (
            format!("Config file deleted successfully in path: {}", file_path),
            exitcode::OK,
        )
    } else {
        if config.is_sensitive_pattern_file_exists() {
            let confirm_message = format!(
                "file {} already exists. do you want to override the existing file?",
                file_path,
            );
            if let Err(e) = promter::confirm(&confirm_message) {
                log::debug!("{:?}", e);
                return Ok(shellclear::CmdExit {
                    code: exitcode::OK,
                    message: None,
                });
            }
        }
        config.init()?;
        (
            format!("Config file created successfully in path: {}", file_path),
            exitcode::OK,
        )
    };

    Ok(shellclear::CmdExit {
        code: exit_code,
        message: Some(message),
    })
}
