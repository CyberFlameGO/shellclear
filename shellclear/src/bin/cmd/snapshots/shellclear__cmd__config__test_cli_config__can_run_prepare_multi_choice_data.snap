---
source: shellclear/src/bin/cmd/config.rs
expression: "get_patter_ignore_multi_choice(&config, &patterns)"
---
(
    [
        "test-1",
        "test-2",
    ],
    [
        "test-3",
    ],
)
